### PR TITLE
Enhancement: Selenium Grid Capability

### PR DESCRIPTION
**Pull request by msghasan**
Signed-off-by: Maimur Hasan <msghasan@gmail.com>

Thanks for contributing to StormCrawler, your efforts are appreciated!

Developer Certificate of Origin
===============================

By contributing to StormCrawler, you accept and agree to the following terms and conditions (the *Developer Certificate of Origin*) for your present and future contributions submitted to StormCrawler. 
Please refer to the *Developer Certificate of Origin* section in [`CONTRIBUTING.md`](CONTRIBUTING.md) for details.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

Before opening a PR, please check that: 

* You've squashed your commits into a single one
* You've described what the PR does or at least point to a related issue
* You've signed-ff your commits with 'git commit -s'
* The code is properly formatted with 'mvn git-code-format:format-code -Dgcf.globPattern=**/*'

Thanks!

### CODE REVIEW

**Comment by msghasan** [comment]
missed a commit

**Review by jnioche — CHANGES_REQUESTED** [review]
Thanks @msghasan! Very kind of you to share your work with the community

A few initial comments
- Formatting a bit off, please run `mvn git-code-format:format-code -Dgcf.globPattern=**/*` prior to committing
- Would be good to explain your approach in a comment in the PR + explain how you tested it
- remove config commented out Elastic/opensearch modules and add it to https://github.com/DigitalPebble/storm-crawler/blob/master/core/src/main/resources/crawler-default.yaml instead 
- remove change to URLFrontier which is unrelated
- Add tests, possibly extending what we recently added in https://github.com/DigitalPebble/storm-crawler/blob/master/core/src/test/java/com/digitalpebble/stormcrawler/protocol/selenium/ProtocolTest.java

**Comment by msghasan** [comment]
@jnioche 
I did all the changes you requested. But for writing a test it is not possible, i think the funcitonality of grid is not available with testcontainers.

https://github.com/testcontainers/testcontainers-java/discussions/5185

**Comment by msghasan** [comment]
Pull Request #1023 Explanation: Enhancing Selenium Grid Capability in Storm Crawler

This pull request aims to improve the functionality of Selenium Grid within the Storm Crawler framework. The implementation involves incorporating a Selenium Grid web address for retrieving remote objects. The key components include two classes: SeleniumGrid.java and its corresponding implementation, SeleniumGridImpl.java.

To manage the number of available browsers in the grid, a LinkedBlockingQueue is utilized. A dynamic check is performed every 5 minutes to ascertain the activity status of the driver object in the queue. We check the session id of the driver ojbect over the grid if it is available we keep the driver else we delete and create new drivers.

The implementation employs a Holder POJO class to store timing information for a driver object and the corresponding driver object added to the queue. In the event of non-session-related exceptions, specifically those arising from website loading issues, the browser object is not discarded; instead, it is reused within the queue.

Testing Approach:

Docker image of Selenium Grid was executed.
The crawler was initiated to assess the proper functioning of the grid and ensure pages were loading successfully.
The page http://gridhost/ui was visited to verify page loading (password: "secret" without quotes).
Grid status was checked at http://gridhost/status.
Through this method, the page source and screenshot of the page were successfully retrieved.
These testing steps were instrumental in validating the effectiveness of the enhanced Selenium Grid functionality in the Storm Crawler.

**Comment by jnioche** [comment]
> @jnioche I did all the changes you requested. But for writing a test it is not possible, i think the funcitonality of grid is not available with testcontainers.
> 
> [testcontainers/testcontainers-java#5185](https://github.com/testcontainers/testcontainers-java/discussions/5185)

I think it does expose the grid. I ran the existing test, called [http://localhost:32779/ui] (the port number might be different) and got the Grid UI. It would have only one node running on it but still, worth using it to test

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            if (!ready) {
+                LOG.warn("Selenium Grid is not ready yet");
+            }
+        }
+        LOG.info("Grid Is Ready to Serve");
+        return (List<Map<String, Object>>) valueMap.get("nodes");
+    }
+
+    private Map<String, Object> getStatusStream() throws IOException {
+        Gson gson = new Gson();
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:70]: Can we use classes from `jackson-databind` here instead of adding a new dependency just for the sake of json parsing?

File: external/urlfrontier/pom.xml [hunk 2]
```diff
@@ -24,6 +24,7 @@
 		<plugins>
 			<plugin>
 				<artifactId>maven-surefire-plugin</artifactId>
+				<version>2.22.2</version>
```
Comment (rzo1) [inline external/urlfrontier/pom.xml:27]: This change seems unrelated to this PR.

File: external/elasticsearch/archetype/src/main/resources/archetype-resources/crawler-conf.yaml [hunk 3]
```diff
@@ -46,6 +46,9 @@ config:
   http.protocol.implementation: "com.digitalpebble.stormcrawler.protocol.okhttp.HttpProtocol"
   https.protocol.implementation: "com.digitalpebble.stormcrawler.protocol.okhttp.HttpProtocol"
 
+  # Please define this for selenium grid configuration
+  selenium.grid.address: http://localhost:4444
```
Comment (rzo1) [inline external/elasticsearch/archetype/src/main/resources/archetype-resources/crawler-conf.yaml:50]: As the change is in `core`, the related adjustment of `crawler-conf.yaml` should also be included in the other archetypes.

File: core/pom.xml [hunk 4]
```diff
@@ -261,7 +263,11 @@
 			<version>${selenium.version}</version>
 			<scope>test</scope>
 		</dependency>
-
+		<dependency>
```
Comment (rzo1) [inline core/pom.xml:266]: see comment below

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.slf4j.LoggerFactory;
+
+public abstract class SeleniumGridProtocol extends AbstractHttpProtocol {
+    protected static final org.slf4j.Logger LOG =
+            LoggerFactory.getLogger(SeleniumGridProtocol.class);
+    protected static LinkedBlockingQueue<Holder> driversQueue;
+    private NavigationFilters filters;
+    private final DesiredCapabilities capabilities = new DesiredCapabilities();
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:41]: This is an unused instance variable. If `capabilities` are defined, we should use them, no?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+    protected String gridAddress;
+
+    @Override
+    public void configure(Config conf) {
+        super.configure(conf);
+        driversQueue = new LinkedBlockingQueue<>();
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:48]: If this is initalized on the same JVM multiple times, we will override the current driver queue. Is this field required to be `static` at all? I would assume, that the grid can deal with multiple clients?

If it needs to be `static`, we should prevent multiple initialization of it (but leads to the same thing, if we are working on multiple JVMs).

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        filters = NavigationFilters.fromConf(conf);
+        gridAddress =
+                ConfUtils.getString(conf, "selenium.grid.address", "http://localhost:4444/wd/hub");
+    }
+
+    protected synchronized List<Map<String, Object>> getAllNodesList() throws IOException {
+        Map<String, Object> valueMap = null;
+        boolean ready = false;
+        while (!ready) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:57]: I am wondering, if we should make this async and use a latch here to prevent active waiting? In any case, we might want to add some sort of wait here to prevent hammering the grid?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            Map<String, Object> map = getStatusStream();
+            valueMap = (Map<String, Object>) map.get("value");
+            ready = (boolean) valueMap.get("ready");
+            if (!ready) {
+                LOG.warn("Selenium Grid is not ready yet");
+            }
+        }
+        LOG.info("Grid Is Ready to Serve");
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:65]: `Grid is ready.`

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+    protected synchronized List<Map<String, Object>> getAllNodesList() throws IOException {
+        Map<String, Object> valueMap = null;
+        boolean ready = false;
+        while (!ready) {
+            Map<String, Object> map = getStatusStream();
+            valueMap = (Map<String, Object>) map.get("value");
+            ready = (boolean) valueMap.get("ready");
+            if (!ready) {
+                LOG.warn("Selenium Grid is not ready yet");
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:62]: Don't think this honours a warning. Maybe also `info` ?

Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:62]: `Selenium Grid is not ready yet.`

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    @Override
+    public void configure(Config conf) {
+        super.configure(conf);
+        driversQueue = new LinkedBlockingQueue<>();
+        filters = NavigationFilters.fromConf(conf);
+        gridAddress =
+                ConfUtils.getString(conf, "selenium.grid.address", "http://localhost:4444/wd/hub");
+    }
+
+    protected synchronized List<Map<String, Object>> getAllNodesList() throws IOException {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:54]: Maybe we can handle a potential exception from `getStatusStream()` inside this method instead of relying on an exception-based code flow?  For example just return `null` or an empty list and deal with it on the caller side?`

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+            for (Map<String, Object> slot : slots) {
+                if (slot.get("session") == null) {
+                    availableSessions++;
+                }
+            }
+        }
+        return availableSessions;
+    }
+
+    public class Holder {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:92]: Should be `static`

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        public RemoteWebDriver driver;
+        public Long time;
+
+        public void setDriver(RemoteWebDriver driver) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:96]: Can be removed as it is never called and driver is set via constructor.

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        updateQueueOfBrowsers();
+    }
+
+    protected void updateQueueOfBrowsers() {
+        timerTask =
+                new TimerTask() {
+                    @Override
+                    public void run() {
+                        try {
+                            List list = getAllNodesList();
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:56]: Can we use a parameterized list here?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    private Timer updateQueue;
+    private int noOfWorkers = 0;
+    private String gridAddress;
+    private final DesiredCapabilities capabilities = new DesiredCapabilities();
+
+    @Override
+    public void configure(Config conf) {
+        super.configure(conf);
+        noOfWorkers = ConfUtils.getInt(conf, "topology.workers", 2);
+        gridAddress = super.gridAddress;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:41]: This is not needed. The grid address is stored in the super class and set via the super.call ?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+public class SeleniumGridImpl extends SeleniumGridProtocol {
+
+    private TimerTask timerTask;
+    private Timer updateQueue;
+    private int noOfWorkers = 0;
+    private String gridAddress;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:34]: can be removed and taken from super class?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+                                int size = list.size();
+                                // Check if the queue size is more than the actual
+                                // no of browsers allowed per worker
+                                // means crawler services are idle because all drivers are in queue
+                                if (driversQueue.size() >= (size / noOfWorkers)) {
+                                    SeleniumGridProtocol.Holder holder = driversQueue.take();
+                                    long totalTime = System.currentTimeMillis() - holder.getTime();
+                                    // clear the queue if the total time spent in the queue is more
+                                    // than 4.5 minutes
+                                    // As after 5 mintues the driver would be idle and will throw
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:68]: `minutes`

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+                                    // exception
+                                    // while we try to use that driver for fetching
+                                    if (totalTime > 1000 * 4.5 * 60) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:71]: 5min is the default of Selenium Grid. Can we make it configurable?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+                                        driversQueue.clear();
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:72]: Why do we clear the full queue? My impression was, that the queue is used to hold `Driver` instances, so we should just remove that specifc driver or did I understand something wrong?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+                                    e);
+                        }
+                    }
+                };
+        // update the queue every 5 minutes
+        updateQueue.schedule(timerTask, 0 * 60 * 1000, 5 * 60 * 1000);
+    }
+
+    protected RemoteWebDriver getDriverFromNode() {
+        int sessionCount = 0;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:101]: unused

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+        RemoteWebDriver driver = null;
+
+        try {
+            LOG.debug("Adding new driver from " + gridAddress);
+            driver = new RemoteWebDriver(new URL(gridAddress), capabilities);
+            WebDriver.Timeouts touts = driver.manage().timeouts();
+            WebDriver.Window window = driver.manage().window();
+            touts.implicitlyWait(Duration.ofSeconds(6));
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:109]: can we add configuration options for these hard-coded values? I think, that we already have them?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+            touts.pageLoadTimeout(Duration.ofSeconds(60));
+            touts.scriptTimeout(Duration.ofSeconds(30));
+            window.setSize(new Dimension(1980, 1280));
+            LOG.debug("Inside getDriverFromGrid to set web drivers" + driver.hashCode());
+        } catch (Exception e) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:114]: We should at least log the exception instead of ignoring it?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+                            }
+                        } catch (Exception e) {
+                            LOG.error(
+                                    "Exception while running task for adding driver to the queue",
+                                    e);
+                        }
+                    }
+                };
+        // update the queue every 5 minutes
+        updateQueue.schedule(timerTask, 0 * 60 * 1000, 5 * 60 * 1000);
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:97]: The first integer value will always be 0.

Can we configure that value?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.TimerTask;
+import org.apache.storm.Config;
+import org.openqa.selenium.Dimension;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+public class SeleniumGridImpl extends SeleniumGridProtocol {
+
+    private TimerTask timerTask;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:31]: Might be a local variable.

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+            if (e.getMessage() != null) {
+                if ((e.getMessage().contains("ERR_NAME_NOT_RESOLVED")
+                        || e.getMessage().contains("ERR_CONNECTION_REFUSED")
+                        || e.getMessage().contains("ERR_CONNECTION_CLOSED")
+                        || e.getMessage().contains("ERR_SSL_PROTOCOL_ERROR")
+                        || e.getMessage().contains("ERR_CONNECTION_RESET")
+                        || e.getMessage().contains("ERR_SSL_VERSION_OR_CIPHER_MISMATCH")
+                        || e.getMessage().contains("ERR_ADDRESS_UNREACHABLE"))) {
+                    LOG.info(
+                            "Exception is of webpage related hence continuing with the driver and adding it back to queue");
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:156]: "Encountered a webpage related exception: '{}'. Adding the driver back to the queue.

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+                } else {
+                    LOG.error(
+                            "Exception wile doing operation via driver url {} with driver hashcode {}"
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:159]: `Exception from driver with url '{}'  and hashcode '{}'. Exception: '{}'`

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+        } catch (Exception e) {
+            return null;
+        }
+    }
+
+    private void closeConnectionGracefully(RemoteWebDriver driver) {
+        try {
+            LOG.info("Before disposing driver : {}", driver.hashCode());
+
+            if (driver.getSessionId() != null) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:187]: Why are we checking twice? If `driver.getSessionId() != null` the `toString()` should be not null as well?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+                if (driver.getSessionId().toString() != null) driver.quit();
+            }
+        } catch (Exception e) {
+            LOG.info("Error while closing driver", e);
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:191]: add driver hash code to the log

**Review by rzo1 — COMMENTED** [review]
Can we implement something similar to `RemoteDriverProtocol` in order to test without running an actual SC crawl?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java (hunk 27 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:191]: sure

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java [hunk 28]
```diff
@@ -0,0 +1,194 @@
...
+            }
+        }
+        LOG.info("Grid Is Ready to Serve");
+        return (List<Map<String, Object>>) valueMap.get("nodes");
+    }
+
+    private Map<String, Object> getStatusStream() throws IOException {
+        Gson gson = new Gson();
+        URL url = new URL(gridAddress + "/status");
+        InputStream stream = url.openStream();
```
[the remark this answers is not in the document]
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:72]: sure

File: external/urlfrontier/pom.xml (hunk 2 above)
Comment (msghasan) [reply external/urlfrontier/pom.xml:27]: will remove it

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java (hunk 5 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:41]: will remove it

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 18 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:72]: Say when the crawler is in a idle mode and has not been running any crawls, the drivers would be sitting  idle and would have lost the session . 

To check every driver session id would be a costly affair. So its better if we clear the queue and again add it from the grid which are active and available in the grid.

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 20 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:109]: sure will do it

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 22 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:97]: sure

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 23 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:31]: done

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 16 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:68]: sharp eyes

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 17 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:71]: the 4.5 minutes you want it to be configurable

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 15 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:34]: done

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 17 above)
Comment (rzo1) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:71]: You can configure the default timeout on the server side of selenium grid, so it would be useful to have a config option in SC to adjust it accordingly.

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 22 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:97]: image

I am naming it 'start.queuing.delay' suggest a property name if you want something different

Comment (rzo1) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:97]: The second value (5min) should most likely also be configurable as it depends on the timeout configured for the grid (default = 5min)

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java (hunk 12 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:96]: Since it is just a pojo class can I move it to some other place, can you please suggest a package where should I move this file. I want to avoid using the new operator inside seleniumgridimpl and directly use the Holder.java as model

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 18 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:72]: Here two sets of list are getting  handled one the grid handles and the other queue we handle per worker. So its better to clear out the queue and try adding it again from the available nodes from the grid.

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 17 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:71]: > You can configure the default timeout on the server side of selenium grid, so it would be useful to have a config option in SC to adjust it accordingly if it is changed on the grid side.

grid.timeout.minutes this is the configuration I am planning to give

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java (hunk 22 above)
Comment (msghasan) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java:97]: made it selenium.grid.delay

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java (hunk 12 above)
Comment (rzo1) [reply core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java:96]: Feel free to make a suggestions.

**Comment by jnioche** [comment]
Happy new year @msghasan! 
Any change you could have a look at the remaining comments from @rzo1? It would be good to include your work in the next release.

**Comment by jnioche** [comment]
Closing because of lack of activity. Can be reopen later if someone wants to pick it up
